src/tools/immutable_object_cache/ObjectCacheStore.cc

```diff
@@ -0,0 +1,244 @@
+      }
+      m_ioctxs.emplace(pool_name, io_ctx);
+    }
+  }
+
+  assert(m_ioctxs.find(pool_name) != m_ioctxs.end());
+
+  librados::IoCtx* ioctx = m_ioctxs[pool_name];
+
+  librados::bufferlist* read_buf = new librados::bufferlist();
+
+  auto ctx = new FunctionContext([this, read_buf, cache_file_name](int ret) {
+      handle_promote_callback(ret, read_buf, cache_file_name);
+   });
+
+   return promote_object(ioctx, object_name, read_buf, ctx);
+}
+
+int ObjectCacheStore::handle_promote_callback(int ret, bufferlist* read_buf,
+  std::string cache_file_name) {
+  ldout(m_cct, 20) << " cache_file_name: " << cache_file_name << dendl;
+
+  // rados read error
+  if(ret != -ENOENT && ret < 0) {
+    lderr(m_cct) << "fail to read from rados" << dendl;
+
+    m_policy->update_status(cache_file_name, OBJ_CACHE_NONE);
+    delete read_buf;
+    return ret;
+  }
+
+  if (ret == -ENOENT) {
+    // object is empty
+    ret = 0;
+  }
+
+  std::string cache_dir = "";
+  uint32_t file_size = ret;
+
+  if (m_dir_num > 0) {
+    auto const pos = cache_file_name.find_last_of('.');
+    cache_dir = "/" + std::to_string(stoul(cache_file_name.substr(pos+1)) % m_dir_num);
+  }
+  // write to cache
+  ObjectCacheFile cache_file(m_cct, cache_dir + "/" + cache_file_name);
+
+  ret = cache_file.write_object_to_file(*read_buf, file_size);
+  if (ret < 0) {
+    lderr(m_cct) << "fail to write cache file" << dendl;
+
+    m_policy->update_status(cache_file_name, OBJ_CACHE_NONE);
+    delete read_buf;
+    return ret;
+  }
+
+  // update metadata
+  assert(OBJ_CACHE_SKIP == m_policy->get_status(cache_file_name));
+  m_policy->update_status(cache_file_name, OBJ_CACHE_PROMOTED);
+  assert(OBJ_CACHE_PROMOTED == m_policy->get_status(cache_file_name));
+
+  delete read_buf;
+  return ret;
+
+  evict_objects();
+}
+
+int ObjectCacheStore::lookup_object(std::string pool_name,
+                                    std::string object_name) {
+  ldout(m_cct, 20) << "object name = " << object_name
+                   << " in pool: " << pool_name << dendl;
+
+  int pret = -1;
+  cache_status_t ret = m_policy->lookup_object(pool_name + object_name);
+
+  switch(ret) {
+    case OBJ_CACHE_NONE: {
+      pret = do_promote(pool_name, object_name);
+      if (pret < 0) {
+        lderr(m_cct) << "fail to start promote" << dendl;
+      }
+      return -1;
+    }
+    case OBJ_CACHE_PROMOTED:
+      return 0;
+    case OBJ_CACHE_SKIP:
+      return -1;
+    default:
+      lderr(m_cct) << "unrecognized object cache status." << dendl;
+      assert(0);
+  }
+}
+
+int ObjectCacheStore::promote_object(librados::IoCtx* ioctx,
+                                     std::string object_name,
+                                     librados::bufferlist* read_buf,
+                                     Context* on_finish) {
+  ldout(m_cct, 20) << "object name = " << object_name << dendl;
+
+  auto ctx = new FunctionContext([on_finish](int ret) {
+    on_finish->complete(ret);
+  });
```

Comment:
it looks like you can just pass `on_finish` to `create_rados_callback` below, without creating this wrapper?

Reply:
done